### DISCUSSION TITLE
[turborepo] Cannot understand what `outputs` in `turbo.json` means.

### DISCUSSION THREAD

**Discussion by gyhyfj** [vercel/turborepo#4568] [Help] [answered]
I've already read the doc, but still cannot understand what does this mean in `turbo.json`, 
it seems no difference in my project with or without this field🙏

**Comment by mehulkar** [comment, marked as answer]
When you configure `outputs` for a task, it tells Turborepo that those files should be cached. For example:
```json
{
  "build": {
    "outputs": ["file.txt"],
  }
}
```

When you run `turbo build`, turbo will expect that a new `file.txt` is created and it will cache that for you. the next time you run `turbo build`, if nothing else has changed, instead of generating that file again, it will use the cached one.

**Reply by pma9** [reply]
@mehulkar Hi so I have this same question and I'm still not getting it. Once we run build, file.txt will be created and if we run build again without changing anything (no workspace dependencies or inputs have changed) then Turborepo automatically has a cache hit and will not rebuild file.txt regardless of whether the `outputs` was set for task. I'm guessing this feature is more for remote caching than running locally?

**Reply by mehulkar** [reply]
> Once we run build, file.txt will be created

correct!

>  if we run build again without changing anything (no workspace dependencies or inputs have changed) then Turborepo automatically has a cache hit

yes!

> will not rebuild file.txt regardless of whether the outputs was set for task.

I think this is the missing piece for you. Before you build again, try _deleting_ `file.txt` manually. Instead of running your `build`, `turbo` will restore `file.txt` from the cache instead of rebuilding. (This cache is stored locally at `./node_modules/.cache/turbo` directory).

Now if you _don't_ delete `file.txt` manually, `turbo` will even skip that restore step, because it knows that `file.txt` hasn't been changed. (There is a background daemon that determines this)

> I'm guessing this feature is more for remote caching than running locally?

There's some truth to this. If you run `npm run build && npm run build`, you are right that there is no need for a cache because your build is already there. But that's not always how it works. Lots of `build` scripts will first _delete_ existing builds so they don't conflict, for example. Or maybe you made a change to your code and then changed it back. After changing it back, you can still get a cache hit and not need to build again. 

Hope this helps.

**Reply by pma9** [reply]
Ahh I see! The switching the code example really nailed it on the head for me. Just for future readers and for my sanity, I'll explain what's going on in my head. So for me, it's a `.next` output that we don't commit to git. I did the following test which shows how this is useful in a local environment.

- Start on branch `abc` and run the build step. turborepo will cache the `.next` directory
- I switch branches to branch `def`, make changes, run the build step which will cause a cache miss due to the change and it will change the `.next` directory
- I then switch back to branch `abc` that I made no changes to. turborepo would ignore that `.next` has changed because its git ignored BUT it will restore the previous correct `.next` from turborepo cache. (without the `outputs` it will not restore and the new changes would still be in `.next` if you use `next start`)

Thanks @mehulkar for taking the time to explain it.
